src/librustc_mir/build/expr/as_rvalue.rs

```diff
@@ -181,6 +181,9 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             ExprKind::Adt {
                 adt_def, variant_index, substs, fields, base
             } => { // see (*) above
+                let is_union = adt_def.adt_kind() == ty::AdtKind::Union;
+                let active_field_index = if is_union { Some(fields[0].name.index()) } else { None };
+
```

Comment:
Future move checker with full support for fragments should accept both `{ let u: Union; u.f = x; u }` and `{ let s: StructWith1Field; s.f = x; s }` I guess, see the last item in https://github.com/rust-lang/rust/pull/36016#issuecomment-242867725.